src/sidebar/components/markdown-editor.js

```diff
@@ -129,74 +133,213 @@ function ToolbarButton({
+  onClick: propTypes.func.isRequired,
   shortcutKey: propTypes.string,
+  tabIndex: propTypes.number.isRequired,
   title: propTypes.string,
 };
 
+/**
+ * @typedef {string} ButtonID
+ * A unique string that can be one of one of:
+ * [bold, italic, quote, link, image, math, numlist, list, preview, help]
+ */
+
+/**
+ * An array of toolbar elements with a roving tab stop. Left and right
+ * array keys can be used to change focus of the elements. Home and end
+ * keys will navigate to the first and last elements respectively.
+ *
+ * Canonical example
+ * https://www.w3.org/TR/wai-aria-practices/examples/toolbar/toolbar.html
+ */
 function Toolbar({ isPreviewing, onCommand, onTogglePreview }) {
+  const buttonIds = {
+    // Ordered buttons
+    bold: 0,
+    italic: 1,
+    quote: 2,
+    link: 3,
+    image: 4,
+    math: 5,
+    numlist: 6,
+    list: 7,
+    help: 8,
+    preview: 9,
+
+    // Total button count
+    maxId: 10,
+  };
+
+  // Keep track of a roving index. The active roving tabIndex
+  // is set to 0, and all other elements are set to -1.
+  const [rovingElement, setRovingElement] = useState(0);
+
+  // An array of refs
+  const buttonRefs = useRef([]).current;
+  if (buttonRefs.length === 0) {
+    // Initialize buttonRefs on first render only
+    for (let i = 0; i <= buttonIds.maxId; i++) {
+      buttonRefs.push(createRef());
+    }
+  }
+
+  /**
+   * Returns the unique button ref associated with the id.
+   *
+   * @param {ButtonID} id
+   * @return {Ref}
+   */
+  const getButtonRef = id => {
+    return buttonRefs[buttonIds[id]];
+  };
+  /**
+   * Sets the element to be both focused and the active roving index.
+   *
+   * @param {number} index - Ordered index that matches the element
+   */
+  const setFocusedElement = index => {
+    setRovingElement(index);
+    buttonRefs[index].current.focus();
+  };
+
+  /**
+   * Handles left and right arrow navigation as well as home and end
+   * keys so the user may navigate the toolbar without multiple tab stops.
+   */
+  const handleKeyDown = e => {
+    let lowerLimit = 0;
+    const upperLimit = buttonIds.maxId - 1;
+    if (isPreviewing) {
+      // When isPreviewing is true, only allow navigation of
+      // the last 2 items.
+      lowerLimit = buttonIds.help;
+    }
+    switch (e.key) {
+      case 'ArrowLeft':
+        if (rovingElement <= lowerLimit) {
+          setFocusedElement(upperLimit);
+        } else {
+          setFocusedElement(rovingElement - 1);
+        }
+        e.preventDefault();
+        break;
+      case 'ArrowRight':
+        if (rovingElement >= upperLimit) {
+          setFocusedElement(lowerLimit);
+        } else {
+          setFocusedElement(rovingElement + 1);
+        }
+        e.preventDefault();
+        break;
+      case 'Home':
+        setFocusedElement(lowerLimit);
+        e.preventDefault();
+        break;
+      case 'End':
+        setFocusedElement(upperLimit);
+        e.preventDefault();
+        break;
```

Comment:
There is a little bit of duplication between the branches here. One pattern I quite like when faced with a long switch or `if/else` chain where each branch executes the same logic but with different values is to do something like the following:

```js
let newFocusedElement = null;
switch (e.key) {
  // Branches set `newFocusedElement` as appropriate.
}
if (newFocusedElement !== null) {
  setFocusedElement(newFocusedElement);
  e.preventDefault();
}
```

The advantage is that it makes it easier to see that each branch is supposed to be doing the same thing and avoids forgetting to put eg. `e.preventDefault()` in one branch, or if you decide to add `e.stopPropagation()` as well, that would now be easier to add.

Reply:
roger that